File: les/milp_problem.cpp
/*
 * Implements milp_problem.hpp interface.
 *
 * Copyright (c) 2012 Oleksander Sviridenko
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied.  See the License for the specific language governing
 * permissions and limitations under the License.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <set>

#include "les/milp_problem.hpp"

const char* MILPP::OBJ_SENSE_TO_STRING[] = {"max", NULL, "min"};

MILPP::MILPP()
{
  set_initial_data();
}

MILPP::MILPP(int nr_cols, int nr_rows)
{
  set_initial_data();
  initialize(nr_cols, nr_rows);
}

MILPP::MILPP(double* c, int nr_cols, double* A, int nr_rows,
             char* s, double* b)
{
  set_initial_data();
  initialize(c, nr_cols, A, nr_rows, s, b);
}

void MILPP::initialize(int nr_cols, int nr_rows)
{
  set_num_cols(nr_cols);
  set_num_rows(nr_rows);
}

void MILPP::initialize(double* c, int nr_cols, double* A, int nr_rows,
                       char* s, double* b)
{
  initialize(nr_cols, nr_rows);
  /* Set vector of coefficients for objective function */
  set_obj_coefs(c);
  set_rows_lower_bounds((double*)calloc(get_num_cols(), sizeof(double)));
  set_rows_upper_bounds(b);
  set_rows_senses(s);
  /* Set constraint matrix */
  set_cons_matrix(A, nr_rows, nr_cols);
}

void MILPP::set_cons_matrix(double* A, int nr_rows, int nr_cols)
{
  initialize(nr_cols, nr_rows);
  /* Fill the constraint matrix  */
  for (size_t i = 0; i < get_num_rows(); i++) {
    for (size_t j = 0; j < get_num_cols(); j++) {
      double v = (A + i * get_num_cols())[j];
      if (!v) continue;
      cons_matrix_.set_coefficient(i, j, v);
      col_to_row_mapping_[j]->insert(i);
      row_to_col_mapping_[i]->insert(j);
    }
  }
  set_rows_lower_bounds((double*)calloc(get_num_cols(), sizeof(double)));
}

void MILPP::set_initial_data()
{
  obj_sense_ = MILPP::OBJ_SENSE_MINIMISATION;
}

void MILPP::set_cons_matrix(const PackedMatrix* matrix)
{
  cons_matrix_ = *matrix;

  for (int i = 0; i < get_num_rows(); i++) {
    PackedVector* row = get_row(i);
    for (int j = 0; j < row->get_num_elements(); j++) {
      col_to_row_mapping_[ row->get_index_by_pos(j) ]->insert(i);
      row_to_col_mapping_[i]->insert( row->get_index_by_pos(j) );
    }
  }
}

void MILPP::dump()
{
  int i;
  int j;
  int offset;
  // Print objective function sense and its coefficients
  for (i = 0; i < get_num_cols(); i++) {
    if (get_obj_coef(i) == 0.0) {
      continue;
    }
    printf("%.1fx%d", get_obj_coef(i), i);
    break;
  }
  for (i = i + 1; i < get_num_cols(); i++) {
    if (get_obj_coef(i) == 0.0) {
      continue;
    }
    printf(" + %.1fx%d", get_obj_coef(i), i);
  }

  printf(" -> %s\n", obj_sense_to_string());
  // Print constraints represented by rows
  printf("subject to\n");
  for (i = 0; i < get_num_rows(); i++) {
    int t = 0; // start from col zero
    PackedVector* row = get_row(i);
    if (!row->size()) {
      continue;
    }
    printf("(%d) ", i);
    for (j = 0; j < row->get_num_elements(); j++) {
      if (j > 0) {
        printf(" + ");
      }
      if ((offset = row->get_index_by_pos(j) - t) >= 1) {
        if (j > 0) {
          offset -= 1;
        }
        offset *= 5 + 3;
        for (int s = 0; s < offset; s++)
          printf(" ");
      }
      t = row->get_index_by_pos(j);

      printf("%.1fx%d", row->get_element_by_pos(j), row->get_index_by_pos(j));
    }
    // Print row sense
    printf(" <= ");
    // Print row upper bound
    printf("%.1f", get_row_upper_bound(i));
    printf("\n");
  }
}

File: les/problems/milp_problem.py
# -*- coding: utf-8; -*-
#
# Copyright (c) 2012-2013 Oleksandr Sviridenko
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
from scipy import sparse

from les.sparse_vector import SparseVector
from les.problems.problem import Problem

class MILPProblem(Problem):
  """This class represents generic MILP problem."""

  def __init__(self, obj_coefs=[], maximaze=True, cons_matrix=None,
               rows_senses=[], rows_upper_bounds=[], rows_lower_bounds=[]):
    self._obj_coefs = None
    self.set_obj_coefs(obj_coefs)
    self._cons_matrix = None
    self._set_cons_matrix(cons_matrix)
    self._rows_senses = rows_senses
    self._rows_upper_bounds = None
    self.set_rows_upper_bounds(rows_upper_bounds)

  def build_subproblem(self, *args, **kwargs):
    return MILPSubproblem(self, *args, **kwargs)

  def get_num_rows(self):
    """Returns the number of rows."""
    return self._cons_matrix.shape[0]

  def get_num_cols(self):
    """Returns the number of columns."""
    return self._cons_matrix.shape[1]

  def _set_cons_matrix(self, m):
    if isinstance(m, np.matrix):
      self._cons_matrix = sparse.csr_matrix(m)
    else:
      self._cons_matrix = m

  def set_obj_coefs(self, coefs):
    if isinstance(coefs, (tuple, list)):
      obj = sparse.dok_matrix((1, len(coefs)), dtype=np.float16)
      for i, c in enumerate(coefs):
        obj[0, i] = c
    else:
      obj = coefs
    self._obj_coefs = SparseVector(obj)

  def get_obj_coefs(self):
    return self._obj_coefs

  def get_rows_upper_bounds(self):
    return self._rows_upper_bounds

  def set_row_upper_bound(self, i, v):
    """Set a single row upper bound."""
    self._rows_upper_bounds[i] = v

  def set_row_bounds(self, lower, upper):
    """Set a single row lower and upper bound."""
    pass

  def set_rows_upper_bounds(self, values):
    """Set rows upper bounds."""
    if isinstance(values, sparse.dok_matrix):
      self._rows_upper_bounds = SparseVector(values)
    elif isinstance(values, (tuple, list)):
      m = sparse.dok_matrix((1, len(values)), dtype=np.float16)
      for i, c in enumerate(values):
        m[0, i] = c
      self._rows_upper_bounds = SparseVector(m)
    elif isinstance(values, SparseVector):
      self._rows_upper_bounds = values
    else:
      raise TypeError()

  def get_cons_matrix(self):
    return self._cons_matrix

  def check_col_solution(self, solution):
    # TODO: solution can be a tuple, list or sparse.base.spmatrix
    if len(solution) != self.get_num_cols():
      raise Exception("%d != %d" % (solution.shape[1], self.get_num_cols()))
    v = self._cons_matrix.dot(solution)
    for i in range(len(v)):
      # TODO: add sense
      if v[i] > self._rows_upper_bounds[i]:
        return False
    return True

class MILPSubproblem(MILPProblem):

  def __init__(self, problem, obj=[], maximaze=True, cons_matrix=None,
               cons_senses=[], upper_bounds=[], shared_cols=[]):
    MILPProblem.__init__(self, obj, maximaze, cons_matrix, cons_senses,
                         upper_bounds)
    self._name = None
    self._problem = problem
    self._shared_cols = shared_cols
    self._local_cols = set(cons_matrix.nonzero()[1]) - shared_cols
    self._deps = dict()

  def __str__(self):
    return "%s[num_shared_cols=%d, num_local_cols=%d]" \
        % (self._name, len(self._shared_cols), len(self._local_cols))

  def _set_name(self, name):
    """Sets subproblem name. Done automatically by solver."""
    if not isinstance(name, str):
      raise TypeError()
    self._name = name

  def get_name(self):
    return self._name

  def get_shared_cols(self):
    return self._shared_cols

  def get_local_cols(self):
    return self._local_cols

  def get_dependencies(self):
    return self._deps

  def add_dependency(self, subproblem, shared_cols):
    if not isinstance(shared_cols, set):
      raise TypeError("shared_cols must be a set")
    if not len(shared_cols):
      raise Exception("No shared cols")
    self._deps[subproblem] = shared_cols
